frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1514 @@
+		pub(crate) dyn_size: u32,
+
+		/// The total size of the migration, over all executions.
+		///
+		/// This only kept around for bookkeeping and debugging.
+		pub(crate) size: u32,
+		/// The total count of top keys in the migration, over all executions.
+		///
+		/// This only kept around for bookkeeping and debugging.
+		pub(crate) top_items: u32,
+		/// The total count of child keys in the migration, over all executions.
+		///
+		/// This only kept around for bookkeeping and debugging.
+		pub(crate) child_items: u32,
+
+		#[codec(skip)]
+		pub(crate) _ph: sp_std::marker::PhantomData<T>,
+	}
+
+	#[cfg(any(feature = "std", feature = "runtime-benchmarks", feature = "try-runtime"))]
+	impl<T: Config> sp_std::fmt::Debug for MigrationTask<T> {
+		fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+			f.debug_struct("MigrationTask")
+				.field(
+					"top",
+					&self.last_top.as_ref().map(|d| sp_core::hexdisplay::HexDisplay::from(d)),
+				)
+				.field(
+					"child",
+					&self.last_child.as_ref().map(|d| sp_core::hexdisplay::HexDisplay::from(d)),
+				)
+				.field("prev_tick_child", &self.prev_tick_child)
+				.field("dyn_top_items", &self.dyn_top_items)
+				.field("dyn_child_items", &self.dyn_child_items)
+				.field("dyn_size", &self.dyn_size)
+				.field("size", &self.size)
+				.field("top_items", &self.top_items)
+				.field("child_items", &self.child_items)
+				.finish()
+		}
+	}
+
+	impl<T: Config> Default for MigrationTask<T> {
+		fn default() -> Self {
+			Self {
+				last_top: Some(Default::default()),
+				last_child: Default::default(),
+				dyn_child_items: Default::default(),
+				dyn_top_items: Default::default(),
+				dyn_size: Default::default(),
+				prev_tick_child: Default::default(),
+				_ph: Default::default(),
+				size: Default::default(),
+				top_items: Default::default(),
+				child_items: Default::default(),
+			}
+		}
+	}
+
+	impl<T: Config> MigrationTask<T> {
+		/// Return true if the task is finished.
+		pub(crate) fn finished(&self) -> bool {
+			self.last_top.is_none() && self.last_child.is_none()
+		}
+
+		/// Returns `true` if the task fully complies with the given limits.
+		pub(crate) fn fully_complies_with(&self, limits: MigrationLimits) -> bool {
+			self.dyn_total_items() <= limits.item && self.dyn_size <= limits.size
+		}
+
+		/// Check if there's any work left, or if we have exhausted the limits already.
+		fn exhausted(&self, limits: MigrationLimits) -> bool {
+			self.last_top.is_none() ||
+				self.dyn_total_items() >= limits.item ||
+				self.dyn_size >= limits.size
+		}
+
+		/// get the total number of keys affected by the current task.
+		pub(crate) fn dyn_total_items(&self) -> u32 {
+			self.dyn_child_items.saturating_add(self.dyn_top_items)
+		}
+
+		/// Migrate keys until either of the given limits are exhausted, or if no more top keys
+		/// exist.
+		///
+		/// Note that this returns after the **first** migration tick that causes exhaustion. In
```

Comment:
I see that there is no issue, but the wording is a bit weird to me, anyway it is fine.
Like this can return without exhausting.
```suggestion
		/// Note that this can return after the **first** migration tick that causes exhaustion. In
```